crates/oxc_transformer/src/es2015/arrow_functions.rs

```diff
@@ -137,22 +137,35 @@ impl<'a> Traverse<'a> for ArrowFunctions<'a> {
             return;
         }
 
-        let ident = match name {
-            JSXElementName::Identifier(ident) => ident,
-            JSXElementName::MemberExpression(member_expr) => {
-                member_expr.get_object_identifier_mut()
+        if let JSXElementName::Identifier(ident) = name {
+            if ident.name == "this" {
+                // We can't produce a proper identifier with a `ReferenceId` because `JSXIdentifier`
+                // lacks that field. https://github.com/oxc-project/oxc/issues/5352
+                // So generate a reference and just use its name.
+                // If JSX transform is enabled, that transform runs before this and will have converted
+                // this to a proper `ThisExpression`, and this visitor won't run.
+                // So only a problem if JSX transform is disabled.
+                let new_ident = self.get_this_name(ctx).create_read_reference(ctx);
+                ident.name = new_ident.name;
+            }
+        }
+    }
+
+    /// Change <this.foo></this.foo> to <_this.foo></_this.foo>, and mark it as found
+    fn enter_jsx_member_expression_object(
+        &mut self,
+        node: &mut JSXMemberExpressionObject<'a>,
+        ctx: &mut TraverseCtx<'a>,
+    ) {
+        if !self.is_inside_arrow_function() {
+            return;
+        }
+
+        if let JSXMemberExpressionObject::IdentifierReference(ident) = node {
+            if ident.name == "this" {
+                let new_ident = self.get_this_name(ctx).create_read_reference(ctx);
+                ident.name = new_ident.name;
```

Comment:
We should also set `reference_id` on the new ident, and handle `<this.foo.bar>`:

```rs
let mut object = node;
let ident = loop {
    match object {
        JSXMemberExpressionObject::IdentifierReference(ident) => {
            if ident.name == "this" {
                break ident;
            }
            return;
        }
        JSXMemberExpressionObject::MemberExpression(member_expr) => {
            object = &mut member_expr.object;
        }
        JSXMemberExpressionObject::Identifier(_) => unreachable!(),
    }
};
let new_ident = self.get_this_name(ctx).create_read_reference(ctx);
ident.name = new_ident.name;
ident.reference_id = new_ident.reference_id;
```